test/integration/job/job_test.go

```diff
@@ -3000,3 +2976,87 @@ func updateJob(ctx context.Context, jobClient typedv1.JobInterface, jobName stri
 	})
 	return job, err
 }
+
+func waitForPodsToBeActive(ctx context.Context, t *testing.T, jobClient typedv1.JobInterface, podCount int32, jobObj *batchv1.Job) {
+	t.Helper()
+	err := wait.PollUntilContextTimeout(ctx, 5*time.Millisecond, wait.ForeverTestTimeout, true, func(context.Context) (done bool, err error) {
+		job, err := jobClient.Get(ctx, jobObj.Name, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		return job.Status.Active == podCount, nil
+	})
+	if err != nil {
+		t.Fatalf("Error waiting for Job pods to become active: %v", err)
+	}
+}
+
+func addFinalizerAndDeletePods(ctx context.Context, t *testing.T, clientSet clientset.Interface, namespace string) {
+	t.Helper()
+	pods, err := clientSet.CoreV1().Pods(namespace).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		t.Fatalf("Failed to list pods: %v", err)
+	}
+	updatePod(ctx, t, clientSet, pods.Items, func(pod *v1.Pod) {
+		pod.Finalizers = append(pod.Finalizers, "fake.example.com/blockDeletion")
```

Comment:
why are you adding a finalizer back?

It looks like you are removing it afterwards.

Reply:
So, I delete pods and assert the job pod status, and this takes some time to assert. After that I think the pods get deleted and the fail pods function errors because the pods are gone, and I want to fail the deleted pods.
I use the finalizer to halt the removal of the pods.

Reply:
Uhm... that's an indication that the job pods have a gracePeriod of zero. You can just change that

Reply:
Ahhh... they are removed immediately because they are not scheduled.
https://github.com/kubernetes/kubernetes/blob/baea2df9d6f02935849e00390e0e8a5fad596f51/pkg/registry/core/pod/strategy.go#L179-L182

So you need the finalizer, but better add it in the job spec, to avoid one update.